File: frontend/src/services/ai.service.ts
/**
 * AI Intelligence Service
 * Handles all AI-powered features including recommendations, QC, anomaly detection, and grouping
 */
import apiClient from './api'
import type {
  ParameterRecommendation,
  PipelineRecommendation,
  RecommendationRequest,
  QCReport,
  AutoQCRequest,
  Anomaly,
  AnomalyDetectionRequest,
  AnomalyDetectionReport,
  SmartGroupingRequest,
  SmartGroupingResult,
  AIAssistantMessage,
  AIAssistantConversation,
  AnalysisInsight,
  InsightsReport,
  ResourcePrediction,
  SuccessPrediction,
  AISystemStatus,
} from '@/types/ai'

class AIService {
  // ============================================================
  // Parameter Recommendations
  // ============================================================

  /**
   * Get pipeline parameter recommendations based on context
   */
  async getParameterRecommendations(request: RecommendationRequest): Promise<PipelineRecommendation> {
    const recommendation = await apiClient.post<PipelineRecommendation>('/ai/recommendations/parameters', request)
    return recommendation
  }

  /**
   * Get recommendations for specific parameter
   */
  async getParameterOptions(
    pipelineType: string,
    parameterName: string,
    context?: Record<string, any>,
  ): Promise<ParameterRecommendation> {
    const recommendation = await apiClient.post<ParameterRecommendation>(
      `/ai/recommendations/parameters/${pipelineType}/${parameterName}`,
      { context },
    )
    return recommendation
  }

  /**
   * Get similar successful pipeline runs
   */
  async getSimilarRuns(request: RecommendationRequest): Promise<
    Array<{
      id: string
      similarity: number
      parameters: Record<string, any>
      outcome: string
    }>
  > {
    const runs = await apiClient.post('/ai/recommendations/similar-runs', request)
    return runs
  }

  // ============================================================
  // Quality Control
  // ============================================================

  /**
   * Run automatic quality control analysis
   */
  async runAutoQC(request: AutoQCRequest): Promise<QCReport> {
    const report = await apiClient.post<QCReport>('/ai/qc/auto-analyze', request)
    return report
  }

  /**
   * Get QC recommendations for a sample
   */
  async getQCRecommendations(sampleId: string): Promise<string[]> {
    const recommendations = await apiClient.get<string[]>(`/ai/qc/recommendations/${sampleId}`)
    return recommendations
  }

  /**
   * Batch QC analysis for multiple samples
   */
  async batchQCAnalysis(sampleIds: string[]): Promise<QCReport[]> {
    const reports = await apiClient.post<QCReport[]>('/ai/qc/batch-analyze', {
      sampleIds,
    })
    return reports
  }

  /**
   * Predict QC issues before sequencing
   */
  async predictQCIssues(
    metadata: Record<string, any>,
  ): Promise<Array<{ issue: string; probability: number; prevention: string }>> {
    const predictions = await apiClient.post('/ai/qc/predict-issues', { metadata })
    return predictions
  }

  // ============================================================
  // Anomaly Detection
  // ============================================================

  /**
   * Detect anomalies in samples or project
   */
  async detectAnomalies(request: AnomalyDetectionRequest): Promise<AnomalyDetectionReport> {
    const report = await apiClient.post<AnomalyDetectionReport>('/ai/anomaly/detect', request)
    return report
  }

  /**
   * Get anomaly details
   */
  async getAnomalyDetails(anomalyId: string): Promise<Anomaly> {
    const anomaly = await apiClient.get<Anomaly>(`/ai/anomaly/${anomalyId}`)
    return anomaly
  }

  /**
   * Apply automatic fix for anomaly
   */
  async applyAnomalyFix(anomalyId: string, fixAction: string): Promise<{ success: boolean; message: string }> {
    const result = await apiClient.post(`/ai/anomaly/${anomalyId}/fix`, {
      action: fixAction,
    })
    return result
  }

  /**
   * Monitor for real-time anomalies
   */
  async monitorAnomalies(projectId: string, _callback: (anomaly: Anomaly) => void): Promise<void> {
    // TODO: Implement WebSocket connection for real-time monitoring
    console.log('Monitoring anomalies for project:', projectId)
    // Placeholder for WebSocket implementation
  }

  // ============================================================
  // Smart Sample Grouping
  // ============================================================

  /**
   * Perform smart sample grouping
   */
  async smartGroupSamples(request: SmartGroupingRequest): Promise<SmartGroupingResult> {
    const result = await apiClient.post<SmartGroupingResult>('/ai/grouping/smart-group', request)
    return result
  }

  /**
   * Suggest comparison groups
   */
  async suggestComparisons(projectId: string): Promise<
    Array<{
      group1: string
      group2: string
      comparisonType: string
      reason: string
      confidence: number
    }>
  > {
    const suggestions = await apiClient.get(`/ai/grouping/suggest-comparisons/${projectId}`)
    return suggestions
  }

  /**
   * Validate sample grouping
   */
  async validateGrouping(groups: Array<{ name: string; sampleIds: string[] }>): Promise<{
    valid: boolean
    issues: string[]
    suggestions: string[]
  }> {
    const validation = await apiClient.post('/ai/grouping/validate', { groups })
    return validation
  }

  // ============================================================
  // AI Assistant
  // ============================================================

  /**
   * Send message to AI assistant
   */
  async sendAssistantMessage(
    conversationId: string,
    message: string,
    context?: Record<string, any>,
  ): Promise<AIAssistantMessage> {
    const response = await apiClient.post<AIAssistantMessage>(
      `/ai/assistant/conversations/${conversationId}/messages`,
      { message, context },
    )
    return response
  }

  /**
   * Create new conversation
   */
  async createConversation(title: string, context?: Record<string, any>): Promise<AIAssistantConversation> {
    const conversation = await apiClient.post<AIAssistantConversation>('/ai/assistant/conversations', {
      title,
      context,
    })
    return conversation
  }

  /**
   * Get conversation history
   */
  async getConversation(conversationId: string): Promise<AIAssistantConversation> {
    const conversation = await apiClient.get<AIAssistantConversation>(`/ai/assistant/conversations/${conversationId}`)
    return conversation
  }

  /**
   * List all conversations
   */
  async listConversations(): Promise<AIAssistantConversation[]> {
    const conversations = await apiClient.get<AIAssistantConversation[]>('/ai/assistant/conversations')
    return conversations
  }

  // ============================================================
  // Analysis Insights
  // ============================================================

  /**
   * Get AI-generated insights for a project
   */
  async getProjectInsights(projectId: string): Promise<InsightsReport> {
    const insights = await apiClient.get<InsightsReport>(`/ai/insights/project/${projectId}`)
    return insights
  }

  /**
   * Get insights for specific analysis
   */
  async getAnalysisInsights(analysisType: string, data: any): Promise<AnalysisInsight[]> {
    const insights = await apiClient.post<AnalysisInsight[]>('/ai/insights/analyze', {
      type: analysisType,
      data,
    })
    return insights
  }

  /**
   * Mark insight as reviewed
   */
  async markInsightReviewed(insightId: string): Promise<void> {
    await apiClient.put(`/ai/insights/${insightId}/reviewed`)
  }

  // ============================================================
  // Predictions
  // ============================================================

  /**
   * Predict resource requirements for pipeline
   */
  async predictResources(
    pipelineType: string,
    sampleCount: number,
    dataSize: number,
    parameters?: Record<string, any>,
  ): Promise<ResourcePrediction> {
    const prediction = await apiClient.post<ResourcePrediction>('/ai/predictions/resources', {
      pipelineType,
      sampleCount,
      dataSize,
      parameters,
    })
    return prediction
  }

  /**
   * Predict success probability for analysis
   */
  async predictSuccess(analysisConfig: Record<string, any>): Promise<SuccessPrediction> {
    const prediction = await apiClient.post<SuccessPrediction>('/ai/predictions/success', analysisConfig)
    return prediction
  }

  /**
   * Predict optimal analysis timeline
   */
  async predictTimeline(projectId: string): Promise<{
    totalDuration: string
    milestones: Array<{ name: string; date: string; confidence: number }>
  }> {
    const timeline = await apiClient.get(`/ai/predictions/timeline/${projectId}`)
    return timeline
  }

  // ============================================================
  // System Status
  // ============================================================

  /**
   * Get AI system status and capabilities
   */
  async getSystemStatus(): Promise<AISystemStatus> {
    const status = await apiClient.get<AISystemStatus>('/ai/status')
    return status
  }

  /**
   * Check if AI features are available
   */
  async isAvailable(): Promise<boolean> {
    try {
      const status = await this.getSystemStatus()
      return status.status === 'operational'
    } catch (error) {
      console.error('Failed to check AI availability:', error)
      return false
    }
  }

  // ============================================================
  // Feedback & Learning
  // ============================================================

  /**
   * Submit feedback on AI recommendation
   */
  async submitFeedback(
    recommendationId: string,
    feedback: {
      helpful: boolean
      accuracy?: number
      comments?: string
    },
  ): Promise<void> {
    await apiClient.post(`/ai/feedback/${recommendationId}`, feedback)
  }

  /**
   * Report incorrect prediction
   */
  async reportIncorrectPrediction(predictionId: string, actualOutcome: any, comments?: string): Promise<void> {
    await apiClient.post(`/ai/feedback/prediction/${predictionId}`, {
      actualOutcome,
      comments,
    })
  }
}

export const aiService = new AIService()
export default aiService

File: backend/app/services/ai_service.py
"""
AI Intelligence Service
Provides AI-powered features including recommendations, QC, anomaly detection, and grouping
"""
from typing import List, Dict, Optional, Any
from datetime import datetime, timedelta
from sqlalchemy.orm import Session
import random
import uuid

from app.schemas.ai import (
    # Enums
    AISystemStatusEnum, QCStatusEnum, AnomalySeverity, AnomalyStatus, InsightCategory,
    # Parameter Recommendations
    RecommendationRequest, ParameterValue, ParameterRecommendation,
    PipelineRecommendation, SimilarRun,
    # Quality Control
    AutoQCRequest, QCMetric, QCRecommendation, QCReport,
    BatchQCRequest, QCIssuePrediction, PredictQCRequest,
    # Anomaly Detection
    AnomalyDetectionRequest, Anomaly, AnomalyDetectionReport,
    AnomalyFixRequest, AnomalyFixResponse,
    # Smart Sample Grouping
    SmartGroupingRequest, SampleGroup, SmartGroupingResult,
    ComparisonSuggestion, GroupValidationRequest, GroupValidationResult,
    # AI Assistant
    AssistantMessageRequest, AIAssistantMessage, AIAssistantConversation,
    CreateConversationRequest,
    # Analysis Insights
    AnalysisInsight, InsightsReport, AnalyzeInsightsRequest,
    # Predictions
    ResourcePredictionRequest, ResourcePrediction, SuccessPrediction,
    TimelineMilestone, TimelinePrediction,
    # System Status
    AICapability, AISystemStatus,
    # Feedback
    FeedbackRequest, IncorrectPredictionRequest, FeedbackResponse,
)


class AIService:
    """
    AI Intelligence Service
    NOTE: This is a mock implementation for frontend integration testing.
    Actual AI/ML models should be integrated in future phases.
    """

    def __init__(self):
        self.conversations_store: Dict[str, AIAssistantConversation] = {}
        self.insights_store: Dict[str, List[AnalysisInsight]] = {}
        self.anomalies_store: Dict[str, Anomaly] = {}

    # ============================================================================
    # Parameter Recommendations
    # ============================================================================

    def get_parameter_recommendations(
        self,
        request: RecommendationRequest,
        db: Session
    ) -> PipelineRecommendation:
        """Get pipeline parameter recommendations based on context"""

        # Mock parameter recommendations
        parameters = []

        if request.pipeline_type == "rna_seq":
            parameters = [
                ParameterValue(
                    name="min_quality",
                    recommended_value=20,
                    confidence=0.92,
                    reasoning="Based on 156 similar successful runs",
                    alternative_values=[15, 25, 30]
                ),
                ParameterValue(
                    name="adapter_sequence",
                    recommended_value="AGATCGGAAGAG",
                    confidence=0.88,
                    reasoning="Most common adapter for Illumina sequencing",
                    alternative_values=["CTGTCTCTTATA", "AGATGTGTATAAGA"]
                ),
                ParameterValue(
                    name="threads",
                    recommended_value=8,
                    confidence=0.95,
                    reasoning="Optimal balance between speed and resource usage",
                    alternative_values=[4, 16, 32]
                ),
            ]
        elif request.pipeline_type == "wgs":
            parameters = [
                ParameterValue(
                    name="min_base_quality",
                    recommended_value=30,
                    confidence=0.90,
                    reasoning="Higher quality threshold for variant calling",
                    alternative_values=[20, 25, 35]
                ),
                ParameterValue(
                    name="min_mapping_quality",
                    recommended_value=20,
                    confidence=0.85,
                    reasoning="Standard mapping quality for WGS",
                    alternative_values=[10, 30, 40]
                ),
            ]
        else:
            parameters = [
                ParameterValue(
                    name="quality_threshold",
                    recommended_value=0.8,
                    confidence=0.80,
                    reasoning="General quality threshold",
                    alternative_values=[0.7, 0.85, 0.9]
                ),
            ]

        return PipelineRecommendation(
            pipeline_type=request.pipeline_type,
            parameters=parameters,
            overall_confidence=0.89,
            similar_runs_analyzed=156,
            estimated_success_rate=0.94,
            estimated_runtime=3600.0 if request.sample_count else None,
            notes=[
                "Recommendations based on similar successful runs",
                "Consider organism-specific parameters for better results"
            ]
        )

    def get_parameter_options(
        self,
        pipeline_type: str,
        parameter_name: str,
        context: Optional[Dict[str, Any]],
        db: Session
    ) -> ParameterRecommendation:
        """Get recommendations for specific parameter"""

        # Mock parameter options based on parameter name
        if parameter_name == "min_quality":
            recommended = 20
            alternatives = [{"value": 15, "pros": "More reads retained"},
                          {"value": 25, "pros": "Higher quality"},
                          {"value": 30, "pros": "Maximum quality"}]
        elif parameter_name == "threads":
            recommended = 8
            alternatives = [{"value": 4, "pros": "Lower resource usage"},
                          {"value": 16, "pros": "Faster processing"},
                          {"value": 32, "pros": "Maximum speed"}]
        else:
            recommended = "auto"
            alternatives = [{"value": "manual", "pros": "More control"}]

        return ParameterRecommendation(
            parameter_name=parameter_name,
            pipeline_type=pipeline_type,
            recommended_value=recommended,
            confidence=0.87,
            reasoning=f"Based on {random.randint(50, 200)} similar runs",
            alternatives=alternatives,
            based_on_runs=random.randint(50, 200)
        )

    def get_similar_runs(
        self,
        request: RecommendationRequest,
        db: Session
    ) -> List[SimilarRun]:
        """Get similar successful pipeline runs"""

        similar_runs = []
        for i in range(5):
            similar_runs.append(SimilarRun(
                id=f"run_{uuid.uuid4().hex[:8]}",
                similarity=round(random.uniform(0.75, 0.95), 2),
                parameters={
                    "min_quality": random.choice([15, 20, 25, 30]),
                    "threads": random.choice([4, 8, 16]),
                    "adapter_sequence": "AGATCGGAAGAG"
                },
                outcome="success",
                success_metric=round(random.uniform(0.85, 0.98), 2)
            ))

        return sorted(similar_runs, key=lambda x: x.similarity, reverse=True)

    # ============================================================================
    # Quality Control
    # ============================================================================

    def run_auto_qc(
        self,
        request: AutoQCRequest,
        db: Session
    ) -> QCReport:
        """Run automatic quality control analysis"""

        # Mock QC metrics
        metrics = [
            QCMetric(
                name="Read Quality",
                value=32.5,
                threshold=30.0,
                status=QCStatusEnum.PASS,
                description="Average Phred quality score"
            ),
            QCMetric(
                name="GC Content",
                value=48.2,
                threshold=None,
                status=QCStatusEnum.PASS,
                description="Percentage of G+C bases"
            ),
            QCMetric(
                name="Duplication Rate",
                value=12.5,
                threshold=15.0,
                status=QCStatusEnum.PASS,
                description="Percentage of duplicate reads"
            ),
            QCMetric(
                name="Adapter Content",
                value=2.1,
                threshold=5.0,
                status=QCStatusEnum.WARNING if random.random() > 0.7 else QCStatusEnum.PASS,
                description="Percentage of reads with adapter sequences"
            ),
        ]

        # Determine overall status
        statuses = [m.status for m in metrics]
        if QCStatusEnum.FAIL in statuses:
            overall_status = QCStatusEnum.FAIL
            overall_score = random.uniform(40, 59)
        elif QCStatusEnum.WARNING in statuses:
            overall_status = QCStatusEnum.WARNING
            overall_score = random.uniform(70, 84)
        else:
            overall_status = QCStatusEnum.PASS
            overall_score = random.uniform(85, 98)

        # Generate recommendations if needed
        recommendations = []
        if overall_status != QCStatusEnum.PASS:
            recommendations.append(QCRecommendation(
                issue="Adapter contamination detected",
                severity="medium",
                recommendation="Run adapter trimming with Trimmomatic or Cutadapt",
                expected_impact="Improve mapping rate by 5-10%"
            ))

        return QCReport(
            sample_id=request.sample_id,
            overall_status=overall_status,
            overall_score=round(overall_score, 1),
            metrics=metrics,
            recommendations=recommendations,
            summary=f"QC analysis completed with {overall_status.value} status. "
                   f"Overall quality score: {overall_score:.1f}/100"
        )

    def get_qc_recommendations(
        self,
        sample_id: str,
        db: Session
    ) -> List[str]:
        """Get QC recommendations for a sample"""

        return [
            "Consider increasing read quality threshold to 25",
            "Remove adapter sequences before alignment",
            "Check for contamination in low-quality regions",
            "Validate library preparation protocol"
        ]

    def batch_qc_analysis(
        self,
        request: BatchQCRequest,
        db: Session
    ) -> List[QCReport]:
        """Batch QC analysis for multiple samples"""

        reports = []
        for sample_id in request.sampleIds:
            report = self.run_auto_qc(
                AutoQCRequest(sample_id=sample_id),
                db
            )
            reports.append(report)

        return reports

    def predict_qc_issues(
        self,
        request: PredictQCRequest,
        db: Session
    ) -> List[QCIssuePrediction]:
        """Predict QC issues before sequencing"""

        predictions = [
            QCIssuePrediction(
                issue="Low yield risk",
                probability=0.23,
                prevention="Verify library concentration before sequencing",
                impact="May result in insufficient coverage"
            ),
            QCIssuePrediction(
                issue="Adapter dimer formation",
                probability=0.15,
                prevention="Perform size selection to remove short fragments",
                impact="Reduces effective sequencing depth"
            ),
        ]

        return predictions

    # ============================================================================
    # Anomaly Detection
    # ============================================================================

    def detect_anomalies(
        self,
        request: AnomalyDetectionRequest,
        db: Session
    ) -> AnomalyDetectionReport:
        """Detect anomalies in samples or project"""

        anomalies = []

        # Generate mock anomalies
        if random.random() > 0.5:
            anomaly = Anomaly(
                type="quality_drop",
                severity=AnomalySeverity.MEDIUM,
                title="Sudden quality drop detected",
                description="Sample quality decreased by 15% compared to recent average",
                affected_resource={"type": "sample", "id": str(uuid.uuid4())},
                metric_values={"quality_score": 72, "average": 85, "deviation": -13},
                suggested_actions=[
                    "Review sample preparation protocol",
                    "Check sequencing instrument calibration",
                    "Compare with control samples"
                ],
                auto_fixable=False,
                confidence=0.87
            )
            anomalies.append(anomaly)
            self.anomalies_store[anomaly.id] = anomaly

        if random.random() > 0.7:
            anomaly = Anomaly(
                type="resource_spike",
                severity=AnomalySeverity.HIGH,
                title="Unusual resource consumption",
                description="Task consumed 3x more memory than similar runs",
                affected_resource={"type": "task", "id": str(uuid.uuid4())},
                metric_values={"memory_gb": 48, "expected": 16, "ratio": 3.0},
                suggested_actions=[
                    "Check for memory leaks in pipeline",
                    "Reduce batch size",
                    "Review input data size"
                ],
                auto_fixable=False,
                confidence=0.92
            )
            anomalies.append(anomaly)
            self.anomalies_store[anomaly.id] = anomaly

        by_severity = {
            "low": 0,
            "medium": 0,
            "high": 0,
            "critical": 0
        }
        for a in anomalies:
            by_severity[a.severity.value] += 1

        return AnomalyDetectionReport(
            project_id=request.project_id,
            anomalies=anomalies,
            total_detected=len(anomalies),
            by_severity=by_severity,
            summary=f"Detected {len(anomalies)} anomalies. "
                   f"{by_severity['high'] + by_severity['critical']} require immediate attention."
        )

    def get_anomaly_details(
        self,
        anomaly_id: str,
        db: Session
    ) -> Anomaly:
        """Get anomaly details"""

        if anomaly_id in self.anomalies_store:
            return self.anomalies_store[anomaly_id]

        # Return mock anomaly if not found
        return Anomaly(
            id=anomaly_id,
            type="unknown",
            severity=AnomalySeverity.LOW,
            title="Anomaly not found",
            description="The requested anomaly details are not available",
            affected_resource={"type": "unknown", "id": "unknown"},
            confidence=0.0
        )

    def apply_anomaly_fix(
        self,
        anomaly_id: str,
        request: AnomalyFixRequest,
        db: Session
    ) -> AnomalyFixResponse:
        """Apply automatic fix for anomaly"""

        # Update anomaly status if exists
        if anomaly_id in self.anomalies_store:
            self.anomalies_store[anomaly_id].status = AnomalyStatus.RESOLVED

        return AnomalyFixResponse(
            success=True,
            message=f"Applied fix: {request.action}",
            anomaly_id=anomaly_id,
            fix_applied=request.action
        )

    # ============================================================================
    # Smart Sample Grouping
    # ============================================================================

    def smart_group_samples(
        self,
        request: SmartGroupingRequest,
        db: Session
    ) -> SmartGroupingResult:
        """Perform smart sample grouping"""

        # Mock sample grouping
        sample_ids = request.sample_ids or [str(uuid.uuid4()) for _ in range(20)]

        groups = []
        samples_per_group = len(sample_ids) // min(request.max_groups, 5)

        for i in range(min(4, request.max_groups)):
            start_idx = i * samples_per_group
            end_idx = start_idx + samples_per_group if i < 3 else len(sample_ids)

            groups.append(SampleGroup(
                name=f"Group {chr(65 + i)}",  # A, B, C, D
                sample_ids=sample_ids[start_idx:end_idx],
                common_attributes={
                    "condition": ["control", "treatment", "vehicle", "drug"][i % 4],
                    "tissue": "liver" if i < 2 else "kidney",
                    "time_point": f"day_{i+1}"
                },
                confidence=round(random.uniform(0.75, 0.95), 2)
            ))

        return SmartGroupingResult(
            groups=groups,
            ungrouped_samples=[],
            grouping_quality=0.87,
            total_samples=len(sample_ids),
            suggestions=[
                "Consider validating group assignments manually",
                "Groups show clear separation by condition",
                "Recommended for differential expression analysis"
            ]
        )

    def suggest_comparisons(
        self,
        project_id: str,
        db: Session
    ) -> List[ComparisonSuggestion]:
        """Suggest comparison groups"""

        return [
            ComparisonSuggestion(
                group1="Control",
                group2="Treatment",
                comparisonType="differential_expression",
                reason="Standard treatment vs control comparison",
                confidence=0.95
            ),
            ComparisonSuggestion(
                group1="Day 1",
                group2="Day 7",
                comparisonType="time_series",
                reason="Temporal progression analysis",
                confidence=0.88
            ),
        ]

    def validate_grouping(
        self,
        request: GroupValidationRequest,
        db: Session
    ) -> GroupValidationResult:
        """Validate sample grouping"""

        issues = []
        suggestions = []

        # Check for common issues
        group_sizes = [len(g.get("sampleIds", [])) for g in request.groups]
        if min(group_sizes) < 3:
            issues.append("Some groups have fewer than 3 samples (minimum for statistical analysis)")
            suggestions.append("Consider merging small groups or collecting more samples")

        if max(group_sizes) > 50:
            suggestions.append("Large groups may benefit from subsampling for computational efficiency")

        return GroupValidationResult(
            valid=len(issues) == 0,
            issues=issues,
            suggestions=suggestions
        )

    # ============================================================================
    # AI Assistant
    # ============================================================================

    def send_assistant_message(
        self,
        conversation_id: str,
        request: AssistantMessageRequest,
        user_id: str,
        db: Session
    ) -> AIAssistantMessage:
        """Send message to AI assistant via the configured provider."""
        from app.services.ai_providers import get_ai_provider

        # Record the user message
        user_msg = AIAssistantMessage(
            conversation_id=conversation_id,
            role="user",
            content=request.message,
            metadata=request.context,
        )

        # Build the message history for the provider
        provider = get_ai_provider()
        history: List[Dict[str, str]] = []

        existing = self.conversations_store.get(conversation_id)
        if existing:
            for m in existing.messages[-20:]:  # last 20 turns
                if m.role in ("user", "assistant"):
                    history.append({"role": m.role, "content": m.content})

        history.append({"role": "user", "content": request.message})

        system_prompt = (
            "You are NGSmodule's AI assistant. Help users with NGS pipeline "
            "configuration, quality control, troubleshooting, and analysis "
            "interpretation. Be concise and cite parameters when relevant."
        )

        try:
            response_text = provider.chat(
                messages=history,
                system=system_prompt,
                temperature=0.4,
                max_tokens=1024,
            )
            metadata = {"provider": provider.name}
        except Exception as exc:
            response_text = (
                "I encountered an issue contacting the AI service. "
                f"Details: {exc}"
            )
            metadata = {"provider": provider.name, "error": str(exc)}

        assistant_msg = AIAssistantMessage(
            conversation_id=conversation_id,
            role="assistant",
            content=response_text,
            metadata=metadata,
        )

        # Persist in the conversation store
        if conversation_id in self.conversations_store:
            self.conversations_store[conversation_id].messages.append(user_msg)
            self.conversations_store[conversation_id].messages.append(assistant_msg)
            self.conversations_store[conversation_id].updated_at = datetime.utcnow()

        return assistant_msg

    def create_conversation(
        self,
        request: CreateConversationRequest,
        user_id: str,
        db: Session
    ) -> AIAssistantConversation:
        """Create new conversation"""

        conversation = AIAssistantConversation(
            title=request.title,
            context=request.context,
            messages=[]
        )

        self.conversations_store[conversation.id] = conversation
        return conversation

    def get_conversation(
        self,
        conversation_id: str,
        user_id: str,
        db: Session
    ) -> AIAssistantConversation:
        """Get conversation history"""

        if conversation_id in self.conversations_store:
            return self.conversations_store[conversation_id]

        # Return mock conversation if not found
        return AIAssistantConversation(
            id=conversation_id,
            title="Conversation not found",
            messages=[]
        )

    def list_conversations(
        self,
        user_id: str,
        db: Session
    ) -> List[AIAssistantConversation]:
        """List all conversations"""

        return list(self.conversations_store.values())

    # ============================================================================
    # Analysis Insights
    # ============================================================================

    def get_project_insights(
        self,
        project_id: str,
        db: Session
    ) -> InsightsReport:
        """Get AI-generated insights for a project"""

        insights = [
            AnalysisInsight(
                category=InsightCategory.QUALITY,
                title="Quality improvement opportunity",
                description="Average quality score is 5% below similar projects",
                importance="medium",
                confidence=0.82,
                actions=[
                    "Review quality control parameters",
                    "Consider re-sequencing low-quality samples"
                ],
                related_data={"current_avg": 82, "expected_avg": 87}
            ),
            AnalysisInsight(
                category=InsightCategory.PERFORMANCE,
                title="Pipeline optimization available",
                description="Tasks are taking 20% longer than expected",
                importance="high",
                confidence=0.89,
                actions=[
                    "Enable parallel processing",
                    "Increase allocated resources"
                ],
                related_data={"avg_runtime": 3600, "expected": 3000}
            ),
        ]

        self.insights_store[project_id] = insights

        by_category = {}
        for insight in insights:
            by_category[insight.category.value] = by_category.get(insight.category.value, 0) + 1

        return InsightsReport(
            project_id=project_id,
            insights=insights,
            summary=f"Generated {len(insights)} actionable insights for project optimization",
            total_insights=len(insights),
            by_category=by_category
        )

    def get_analysis_insights(
        self,
        request: AnalyzeInsightsRequest,
        db: Session
    ) -> List[AnalysisInsight]:
        """Get insights for specific analysis"""

        return [
            AnalysisInsight(
                category=InsightCategory.OPTIMIZATION,
                title=f"Optimization suggestion for {request.type}",
                description="AI-detected optimization opportunity",
                importance="medium",
                confidence=0.78,
                actions=["Review analysis parameters"]
            )
        ]

    def mark_insight_reviewed(
        self,
        insight_id: str,
        project_id: str,
        db: Session
    ) -> None:
        """Mark insight as reviewed"""

        if project_id in self.insights_store:
            for insight in self.insights_store[project_id]:
                if insight.id == insight_id:
                    insight.reviewed = True
                    break

    # ============================================================================
    # Predictions
    # ============================================================================

    def predict_resources(
        self,
        request: ResourcePredictionRequest,
        db: Session
    ) -> ResourcePrediction:
        """Predict resource requirements for pipeline"""

        # Simple mock prediction based on pipeline type
        base_cpu = 4
        base_memory = 8.0
        base_storage = 50.0
        base_duration = 1800.0

        # Scale by sample count and data size
        cpu_cores = base_cpu * (1 + request.sampleCount // 10)
        memory_gb = base_memory * (1 + request.dataSize / 100)
        storage_gb = base_storage * request.sampleCount
        duration = base_duration * request.sampleCount

        return ResourcePrediction(
            cpu_cores=min(cpu_cores, 64),
            memory_gb=round(min(memory_gb, 256.0), 1),
            storage_gb=round(storage_gb, 1),
            estimated_duration_seconds=duration,
            confidence=0.85,
            notes=[
                "Estimates based on historical data",
                "Actual usage may vary based on data characteristics"
            ]
        )

    def predict_success(
        self,
        analysis_config: Dict[str, Any],
        db: Session
    ) -> SuccessPrediction:
        """Predict success probability for analysis"""

        # Mock success prediction
        success_prob = random.uniform(0.75, 0.95)

        risk_factors = []
        success_factors = []

        if success_prob < 0.85:
            risk_factors.append("Quality scores below recommended threshold")
            risk_factors.append("Limited computational resources")
        else:
            success_factors.append("High-quality input data")
            success_factors.append("Optimal parameter configuration")

        return SuccessPrediction(
            success_probability=round(success_prob, 2),
            risk_factors=risk_factors,
            success_factors=success_factors,
            recommendations=[
                "Validate input data quality before proceeding",
                "Ensure sufficient computational resources are available"
            ],
            confidence=0.82
        )

    def predict_timeline(
        self,
        project_id: str,
        db: Session
    ) -> TimelinePrediction:
        """Predict optimal analysis timeline"""

        now = datetime.utcnow()
        milestones = [
            TimelineMilestone(
                name="QC Complete",
                date=(now + timedelta(days=2)).isoformat(),
                confidence=0.90
            ),
            TimelineMilestone(
                name="Alignment Complete",
                date=(now + timedelta(days=5)).isoformat(),
                confidence=0.85
            ),
            TimelineMilestone(
                name="Analysis Complete",
                date=(now + timedelta(days=10)).isoformat(),
                confidence=0.78
            ),
        ]

        return TimelinePrediction(
            totalDuration="10 days",
            milestones=milestones,
            confidence=0.82
        )

    # ============================================================================
    # System Status
    # ============================================================================

    def get_system_status(self, db: Session) -> AISystemStatus:
        """Get AI system status and capabilities"""

        capabilities = [
            AICapability(
                name="Parameter Recommendations",
                available=True,
                version="1.0.0",
                description="AI-powered pipeline parameter suggestions"
            ),
            AICapability(
                name="Auto QC",
                available=True,
                version="1.0.0",
                description="Automatic quality control analysis"
            ),
            AICapability(
                name="Anomaly Detection",
                available=True,
                version="1.0.0",
                description="Real-time anomaly detection in pipelines"
            ),
            AICapability(
                name="Smart Grouping",
                available=True,
                version="1.0.0",
                description="Intelligent sample grouping and comparison suggestions"
            ),
            AICapability(
                name="AI Assistant",
                available=True,
                version="1.0.0",
                description="Interactive AI assistant for guidance"
            ),
            AICapability(
                name="Resource Prediction",
                available=True,
                version="1.0.0",
                description="Predictive resource requirement estimation"
            ),
        ]

        return AISystemStatus(
            status=AISystemStatusEnum.OPERATIONAL,
            version="1.0.0",
            capabilities=capabilities,
            models_loaded=["mock_recommendation_model", "mock_qc_model", "mock_anomaly_model"],
            uptime_seconds=86400.0,
            notes=["Mock implementation for frontend integration testing"]
        )

    # ============================================================================
    # Feedback & Learning
    # ============================================================================

    def submit_feedback(
        self,
        recommendation_id: str,
        feedback: FeedbackRequest,
        user_id: str,
        db: Session
    ) -> FeedbackResponse:
        """Submit feedback on AI recommendation"""

        return FeedbackResponse(
            success=True,
            message="Thank you for your feedback! It helps improve our AI models."
        )

    def report_incorrect_prediction(
        self,
        prediction_id: str,
        report: IncorrectPredictionRequest,
        user_id: str,
        db: Session
    ) -> FeedbackResponse:
        """Report incorrect prediction"""

        return FeedbackResponse(
            success=True,
            message="Prediction correction recorded. Our models will learn from this feedback."
        )


# Create singleton instance
ai_service = AIService()
